Mono.Addins.Setup/Mono.Addins.Setup/AddinStore.cs

```diff
@@ -673,10 +676,22 @@ internal string DownloadFile (IProgressMonitor monitor, string url)
 				throw;
 			} finally {
 				monitor.EndTask ();
-				monitor.EndTask ();
 			}
 		}
-			
+
+		static bool WaitForTask (Task<DownloadFileRequest> task, IProgressMonitor monitor)
+		{
+			bool result = SpinWait.SpinUntil (() => {
+				return monitor.IsCancelRequested || task.IsCompleted || task.IsFaulted;
+			}, 100000); // Use same default timeout as HttpClient.
```

Comment:
Will this spin wait for 100s?

Reply:
It checks the result from the delegate and will only wait 100s if the delegate returns false.

Reply:
From my reading of the documentation, it does busy waiting while it can, then yields to a different thread:
https://source.dot.net/#System.Private.CoreLib/shared/System/Threading/SpinWait.cs,339

That means we'll spin the CPU a lot waiting for something to happen.

Reply:
Can we not signal completion in a different way and use a manual reset event?

Reply:
ie. chain a continuation on the main task that toggles the manualresetevent and do a blocking wait on thsi thread rather than spinning?

Reply:
Same, register a callback on the CancellationToken to trigger the cancellation int hs code stack.

Reply:
We could just do `task.Result` here instead - assuming we have no problem with code that does that. Need to check what the UI thread is doing here.

Reply:
Can't we async up the stack?

Reply:
`task.Wait (cancellationToken)` is probably a more CPU friendly version of `SpinWait.SpinUntil` in this context. Web requests are probably slow enough that a regular Wait would be more beneficial than a SpinWait :)

Both versions should have an identical risk of deadlocking so they should be interchangeable.

Reply:
Maybe we should have 2 variants, one sync, one async for the API?

Reply:
Does not seem to be a simple way to do this. I added a CancellationToken to the ProgressStatusMonitor but its Cancel method will never be called. Most of the logic for cancellation is done with the IProgressStatus derived classes and its Cancel and IsCanceled method/property, which is implemented by various UI dialogs. For now I think I will change it to use `task.Wait ()`. The task.Wait will be run on a background thread when run with MonoDevelop.